skills/types/bool_file.go

```diff
@@ -0,0 +1,140 @@
+	// The value of standard exported GPIO
+	boolFileGPIOValuePattern,
+}
+
+// SanitizeSkill checks and possibly modifies a skill.
+// Valid "bool-file" skills must contain the attribute "path".
+func (t *BoolFileType) SanitizeSkill(skill *skills.Skill) error {
+	if t.Name() != skill.Type {
+		panic(fmt.Sprintf("skill is not of type %q", t))
+	}
+	path, ok := skill.Attrs["path"].(string)
+	if !ok || path == "" {
+		return fmt.Errorf("bool-file must contain the path attribute")
+	}
+	if strings.Contains(path, "..") {
+		return fmt.Errorf("bool-file path cannot contain %q", "..")
+	}
+	for _, pattern := range boolFileAllowedPathPatterns {
+		if pattern.MatchString(path) {
+			return nil
+		}
+	}
+	return fmt.Errorf("bool-file can only point at LED brightness or GPIO value")
+}
+
+// SanitizeSlot checks and possibly modifies a skill slot.
+func (t *BoolFileType) SanitizeSlot(skill *skills.Slot) error {
+	if t.Name() != skill.Type {
+		panic(fmt.Sprintf("skill slot is not of type %q", t))
+	}
+	// NOTE: currently we don't check anything on the slot side.
+	return nil
+}
+
+// SkillSecuritySnippet returns the configuration snippet required to provide a bool-file skill.
+// Producers gain control over exporting, importing GPIOs as well as
+// controlling the direction of particular pins.
+func (t *BoolFileType) SkillSecuritySnippet(skill *skills.Skill, securitySystem skills.SecuritySystem) ([]byte, error) {
+	gpioSnippet := []byte(`
+/sys/class/gpio/export rw,
+/sys/class/gpio/unexport rw,
+/sys/class/gpio/gpio[0-9]+/direction rw,
+`)
+	switch securitySystem {
+	case skills.SecurityApparmor:
+		// To provide GPIOs we need extra permissions to export/unexport and to
+		// set the direction of each pin.
+		if t.isGPIO(skill) {
+			return gpioSnippet, nil
+		}
+		return nil, nil
+	case skills.SecuritySeccomp, skills.SecurityDBus:
+		return nil, nil
+	default:
+		return nil, skills.ErrUnknownSecurity
+	}
+}
+
+// SlotSecuritySnippet returns the configuration snippet required to use a bool-file skill.
+// Consumers gain permission to read, write and lock the designated file.
+func (t *BoolFileType) SlotSecuritySnippet(skill *skills.Skill, securitySystem skills.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case skills.SecurityApparmor:
+		// Allow write and lock on the file designated by the path.
+		// Dereference symbolic links to file path handed out to apparmor since
+		// sysfs is full of symlinks and apparmor requires uses real path for
+		// filtering.
+		path, err := t.dereferencedPath(skill)
+		if err != nil {
+			return nil, fmt.Errorf("cannot compute skill slot security snippet: %v", err)
+		}
+		return []byte(fmt.Sprintf("%s rwk,\n", path)), nil
+	case skills.SecuritySeccomp, skills.SecurityDBus:
+		return nil, nil
+	default:
+		return nil, skills.ErrUnknownSecurity
+	}
+}
+
+func (t *BoolFileType) dereferencedPath(skill *skills.Skill) (string, error) {
+	if path, ok := skill.Attrs["path"].(string); ok {
```

Comment:
As far as I can tell here `skill.Attrs["path"]` can come from "outside" (as `skill` is passed in unchanged from an exported function), so you need `filepath.Clean` here again, don't you?